sdk/secp256r1/src/lib.rs

```diff
@@ -0,0 +1,604 @@
+//! Instructions for the [secp256r1 native program][np].
+//!
+//! [np]: https://docs.solana.com/developing/runtime-facilities/programs#secp256r1-program
+
+#[cfg(not(target_arch = "wasm32"))]
+use {
+    bytemuck::bytes_of,
+    openssl::{
+        bn::{BigNum, BigNumContext},
+        ec::{EcGroup, EcKey, EcPoint},
+        ecdsa::EcdsaSig,
+        nid::Nid,
+        pkey::PKey,
+        sign::{Signer, Verifier},
+    },
+    solana_instruction::Instruction,
+};
+use {
+    bytemuck::{Pod, Zeroable},
+    solana_feature_set::FeatureSet,
+    solana_precompile_error::PrecompileError,
+};
+
+solana_pubkey::declare_id!("Secp256r1SigVerify1111111111111111111111111");
+
+#[cfg(not(target_arch = "wasm32"))]
+pub const COMPRESSED_PUBKEY_SERIALIZED_SIZE: usize = 33;
+#[cfg(not(target_arch = "wasm32"))]
+pub const SIGNATURE_SERIALIZED_SIZE: usize = 64;
+#[cfg(not(target_arch = "wasm32"))]
+pub const SIGNATURE_OFFSETS_SERIALIZED_SIZE: usize = 14;
+// bytemuck requires structures to be aligned
+#[cfg(not(target_arch = "wasm32"))]
+pub const SIGNATURE_OFFSETS_START: usize = 2;
+#[cfg(not(target_arch = "wasm32"))]
+pub const DATA_START: usize = SIGNATURE_OFFSETS_SERIALIZED_SIZE + SIGNATURE_OFFSETS_START;
+
+// Order as defined in SEC2: 2.7.2 Recommended Parameters secp256r1
+#[cfg(not(target_arch = "wasm32"))]
+pub const SECP256R1_ORDER: [u8; 32] = [
+    0xFF, 0xFF, 0xFF, 0xFF, 0x00, 0x00, 0x00, 0x00, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+    0xBC, 0xE6, 0xFA, 0xAD, 0xA7, 0x17, 0x9E, 0x84, 0xF3, 0xB9, 0xCA, 0xC2, 0xFC, 0x63, 0x25, 0x51,
+];
+
+// Computed SECP256R1_ORDER - 1
+#[cfg(not(target_arch = "wasm32"))]
+pub const SECP256R1_ORDER_MINUS_ONE: [u8; 32] = [
+    0xFF, 0xFF, 0xFF, 0xFF, 0x00, 0x00, 0x00, 0x00, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+    0xBC, 0xE6, 0xFA, 0xAD, 0xA7, 0x17, 0x9E, 0x84, 0xF3, 0xB9, 0xCA, 0xC2, 0xFC, 0x63, 0x25, 0x50,
+];
+// Computed half order
+#[cfg(not(target_arch = "wasm32"))]
+const SECP256R1_HALF_ORDER: [u8; 32] = [
+    0x7F, 0xFF, 0xFF, 0xFF, 0x80, 0x00, 0x00, 0x00, 0x7F, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+    0xDE, 0x73, 0x7D, 0x56, 0xD3, 0x8B, 0xCF, 0x42, 0x79, 0xDC, 0xE5, 0x61, 0x7E, 0x31, 0x92, 0xA8,
+];
+
+#[derive(Default, Debug, Copy, Clone, Zeroable, Pod, Eq, PartialEq)]
+#[repr(C)]
+pub struct Secp256r1SignatureOffsets {
+    signature_offset: u16, // offset to compact secp256r1 signature of 64 bytes
+    signature_instruction_index: u16, // instruction index to find signature
+    public_key_offset: u16, // offset to compressed public key of 33 bytes
+    public_key_instruction_index: u16, // instruction index to find public key
+    message_data_offset: u16, // offset to start of message data
+    message_data_size: u16, // size of message data
+    message_instruction_index: u16, // index of instruction data to get message data
+}
+
+#[cfg(not(target_arch = "wasm32"))]
```

Comment:
There's obvious a ton of these `#[cfg(not(target_arch = "wasm32"))]` decorating everything in this module. I wonder if using another sub-module here might make things easier to maintain?

```rust
#[cfg(not(target_arch = "wasm32"))]
mod not_for_wasm {
    // Basically everything from this file except `Secp256r1SignatureOffsets`.
}

#[cfg(not(target_arch = "wasm32"))]
pub use not_for_wasm::verify;

#[cfg(target_arch = "wasm32")]
pub fn verify(
    _data: &[u8],
    _instruction_datas: &[&[u8]],
    _feature_set: &FeatureSet,
) -> Result<(), PrecompileError> {
    Err(PrecompileError::InvalidSignature)
}
```

Not really sure if that's better, though. Lmk what you think!